### ISSUE TITLE
RFC: Three-API Architecture for docker-wrapper

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/docker-wrapper#73]
## Overview

Design a three-API architecture for docker-wrapper that provides different interfaces while sharing backend infrastructure:

1. **API 1: CLI Binary** - Wraps `docker` command (current implementation)
2. **API 2: CLI-style API Backend** - Same interface as API 1, but calls dockerd REST API directly  
3. **API 3: Raw API** - Direct API mapping, bollard competitor with better ergonomics

### API 1: CLI Binary (Existing)
```rust
// Wraps docker command exactly
let docker = Docker::new();
let result = docker.run("nginx:latest")
    .name("web")
    .port("8080:80")
    .detach()
    .execute()
    .await?;
```

**Characteristics:**
- Subprocess calls to `docker` binary
- Inherits Docker CLI remote capabilities (contexts, SSH)
- Most stable/reliable (Docker handles edge cases)
- Current implementation

### API 2: CLI-style API Backend
```rust
// Identical interface to API 1, different backend
let docker = Docker::with_api_backend()?;
let result = docker.run("nginx:latest")  // Same API!
    .name("web")
    .port("8080:80") 
    .detach()
    .execute()
    .await?;
```

**Characteristics:**
- Same builder pattern interface as API 1
- Direct REST API calls to dockerd
- Better performance (no subprocess overhead)
- Remote connections via direct API endpoints

### API 3: Raw API
```rust
// Different interface, direct API mapping
let docker = DockerApi::new()?;
let container = docker.containers()
    .create("nginx:latest")
    .name("web")
    .port(8080, 80)
    .detached()
    .execute()
    .await?;

container.start().await?;
```

**Characteristics:**
- Handle-based operations
- Structured types (not string parsing)
- Direct dockerd REST API mapping
- Bollard competitor with better ergonomics

### Shared Backend Trait
```rust
#[async_trait]
pub trait DockerBackend: Send + Sync {
    async fn create_container(&self, config: ContainerConfig) -> Result<String, DockerError>;
    async fn start_container(&self, id: &str) -> Result<(), DockerError>;
    async fn list_containers(&self, filters: ContainerFilters) -> Result<Vec<Container>, DockerError>;
    async fn stop_container(&self, id: &str, timeout: Option<u64>) -> Result<(), DockerError>;
    
    fn backend_type(&self) -> BackendType;
    fn supports_remote(&self) -> bool;
    fn supports_streaming(&self) -> bool;
    fn supports_compose(&self) -> bool;
    fn supports_buildx(&self) -> bool;
}
```

### Easy Commands (Single REST call)
- `docker ps` → `GET /containers/json`
- `docker stop <id>` → `POST /containers/{id}/stop`
- `docker rm <id>` → `DELETE /containers/{id}`
- `docker images` → `GET /images/json`

### Medium Commands (Multiple calls or complex JSON)
- `docker run` → `POST /containers/create` + `POST /containers/{id}/start`
- `docker create` → `POST /containers/create` (complex JSON body)

### Hard Commands (Streaming)
- `docker logs` → `GET /containers/{id}/logs` (streaming response)
- `docker pull` → `POST /images/create` (streaming with progress)

### Impossible Commands (CLI only)
- `docker compose *` → No REST API, must use CLI backend
- `docker buildx *` → CLI plugin only

### Phase 1: Backend Abstraction (2-3 weeks)
1. Define `DockerBackend` trait
2. Refactor existing CLI code to implement trait
3. Create shared `ContainerConfig` types
4. Add capability detection methods

### Phase 2: API Backend Implementation (4-6 weeks)
1. Build custom REST client for dockerd
2. Implement `ApiBackend` with core commands
3. Map CLI-style options to API calls
4. Handle Unix sockets, named pipes, TCP connections
5. Add error mapping between API responses and CLI-style errors

### Phase 3: User-Facing APIs (6-8 weeks)
1. API 1 & 2: Same interface, different backends
2. API 3: New interface for direct API access
3. Backend selection logic
4. Feature flag configuration

## Feature Gating

```toml
[features]
default = ["cli-binary"]

# API Options
cli-binary = []     # API 1: Wrap docker binary
cli-api = []        # API 2: CLI-style interface, API backend  
raw-api = []        # API 3: Raw API interface

# Backend Support
api-client = []     # Custom REST client for cli-api and raw-api
```

## Key Benefits

1. **Progressive Enhancement**: Start stable (API 1) → add performance (API 2) → add power (API 3)
2. **Zero Breaking Changes**: Each API evolves independently
3. **Market Coverage**: Different APIs serve different user segments
4. **Competitive Advantage**: Unique positioning in Rust Docker ecosystem
5. **Future-Proof**: Can adapt as Docker ecosystem evolves

## Remote Docker Support

**CLI Binary (API 1):**
- Docker contexts: `docker --context remote ps`
- Environment variables: `DOCKER_HOST=ssh://user@host`
- SSH forwarding: Docker CLI handles everything

**API Backend (APIs 2&3):**
- Direct TCP connections: `http://remote-host:2376`
- TLS authentication
- Custom endpoint configuration

### Technical Risks
- **Scope Creep**: Three APIs could become maintenance burden
- **API Parity**: Keeping all three APIs in sync
- **Testing Complexity**: Each backend needs full test coverage
- **Documentation**: Three sets of docs/examples

### Mitigation Strategy
- Start with Phase 1 (backend abstraction)
- Validate with core commands first
- Get user feedback before expanding
- Consider focusing on API 1+2 initially

## User Segments

| API | Target Users | Use Cases |
|-----|-------------|-----------|
| API 1 | Shell script replacements, existing users | Simple automation, CI/CD |
| API 2 | Performance-sensitive apps | Cloud deployments, high-throughput |
| API 3 | Advanced users | Direct Docker integration, complex workflows |

## Success Metrics

- **Phase 1**: Backend abstraction with no breaking changes
- **Phase 2**: 50% performance improvement for API 2 vs API 1
- **Phase 3**: Feature parity with bollard for core operations

## Alternative Approach

Consider starting smaller:
```rust
// Same interface, different backends
let docker = Docker::new()
    .with_backend(BackendType::Api)? // Optional feature
    .run("nginx")
    .execute().await?;
```

This provides immediate performance benefits while maintaining a manageable scope.

## Discussion Points

1. Is the three-API approach too complex?
2. Should we start with just API 1+2 and defer API 3?
3. How do we handle feature differences between backends?
4. What's the priority of this vs. other roadmap items?

This is an RFC (Request for Comments) - feedback and discussion welcome!
